File: PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/logging/GamLogUtil.java
/*
 *    Copyright 2018-2019 Prebid.org, Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package org.prebid.mobile.logging;

import android.util.Log;

import androidx.annotation.Size;

public class GamLogUtil {
    private static final String GAM_TAG = "GAM";

    public static final int NONE = -1;
    public static final int VERBOSE = Log.VERBOSE; // 2
    public static final int DEBUG = Log.DEBUG; // 3
    public static final int INFO = Log.INFO; // 4
    public static final int WARN = Log.WARN; // 5
    public static final int ERROR = Log.ERROR; // 6
    public static final int ASSERT = Log.ASSERT; // 7

    private static int logLevel;

    private GamLogUtil() {
    }

    public static void setLogLevel(int level) {
        logLevel = level;
    }

    public static int getLogLevel() {
        return logLevel;
    }

    /**
     * Configure log server sending
     *
     * @param serverUrl URL of the log server endpoint
     * @param enabled   Whether to enable sending logs to server
     */
    public static void configureLogServer(String serverUrl, boolean enabled) {
        LogServerSender.getInstance().configure(serverUrl, enabled);
    }

    /**
     * Prints a message with VERBOSE priority and default GAM_TAG
     */
    public static void verbose(String message) {
        verbose(GAM_TAG, message);
    }

    /**
     * Prints a message with DEBUG priority and default GAM_TAG
     */
    public static void debug(String message) {
        debug(GAM_TAG, message);
    }

    /**
     * Prints a message with INFO priority and default GAM_TAG
     */
    public static void info(String message) {
        info(GAM_TAG, message);
    }

    /**
     * Prints a message with WARNING priority and default GAM_TAG
     */
    public static void warning(String message) {
        warning(GAM_TAG, message);
    }

    /**
     * Prints a message with ERROR priority and default GAM_TAG
     */
    public static void error(String message) {
        error(GAM_TAG, message);
    }

    /**
     * Prints a message with VERBOSE priority.
     */
    public static void verbose(@Size(max = 23) String tag, String msg) {
        print(VERBOSE, tag, msg);
    }

    /**
     * Prints a message with DEBUG priority.
     */
    public static void debug(@Size(max = 23) String tag, String msg) {
        print(DEBUG, tag, msg);
    }

    /**
     * Prints a message with INFO priority.
     */
    public static void info(@Size(max = 23) String tag, String msg) {
        print(INFO, tag, msg);
    }

    /**
     * Prints a message with WARN priority.
     */
    public static void warning(@Size(max = 23) String tag, String msg) {
        print(WARN, tag, msg);
    }

    /**
     * Prints a message with ERROR priority.
     */
    public static void error(@Size(max = 23) String tag, String msg) {
        print(ERROR, tag, msg);
    }

    /**
     * Prints a message with ASSERT priority.
     */
    public static void wtf(@Size(max = 23) String tag, String msg) {
        print(ASSERT, tag, msg);
    }

    /**
     * Prints a message with ERROR priority and exception.
     */
    public static void error(final String tag, String message, Throwable throwable) {
        if (tag == null || message == null) {
            return;
        }
        Log.e(getTagWithBase(tag), message, throwable);
    }

    /**
     * Prints information with set priority. Every tag
     */
    private static void print(int messagePriority, String tag, String message) {
        if (tag == null || message == null) {
            return;
        }
        String finalTag = getTagWithBase(tag);
        Log.println(messagePriority, finalTag, message);

        // Send to server
        LogServerSender.getInstance().sendLog(messagePriority, finalTag, message);
    }

    /**
     * Helper method to add Prebid tag to logging messages.
     */
    private static String getTagWithBase(String tag) {
        StringBuilder result = new StringBuilder();

        String prefix = "GAM";
        if (tag.startsWith(prefix)) {
            result.append(tag);
        } else {
            result.append(prefix).append(tag);
        }

        if (result.length() > 23) {
            return result.substring(0, 22);
        } else {
            return result.toString();
        }
    }
}

File: PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/logging/GamLogUtil.kt
package org.prebid.mobile.logging

import android.util.Log
import androidx.annotation.Size
import org.prebid.mobile.api.data.AdFormat

object GamLogUtil {
    private const val GAM_TAG = "GAM"

    const val NONE = -1
    const val VERBOSE = Log.VERBOSE // 2
    const val DEBUG = Log.DEBUG // 3
    const val INFO = Log.INFO // 4
    const val WARN = Log.WARN // 5
    const val ERROR = Log.ERROR // 6
    const val ASSERT = Log.ASSERT // 7

    @JvmStatic
    var logLevel: Int = 0

    /**
     * Configure log server sending
     *
     * @param serverUrl URL of the log server endpoint
     * @param enabled   Whether to enable sending logs to server
     */
    @JvmStatic
    fun configureLogServer(serverUrl: String?, enabled: Boolean) {
        LogServerSender.getInstance().configure(serverUrl, enabled)
    }

    /**
     * Prints a message with INFO priority and default GAM_TAG
     */
    @JvmStatic
    fun info(message: String, gamStatus: GamStatus, adFormat: AdFormat, adUnitId: String) {
        info(GAM_TAG, message, gamStatus, adFormat, adUnitId)
    }

    /**
     * Prints a message with ERROR priority and default GAM_TAG
     */
    @JvmStatic
    fun error(message: String, adFormat: AdFormat, adUnitId: String) {
        error(GAM_TAG, message, adFormat, adUnitId)
    }

    /**
     * Prints a message with INFO priority.
     */
    @JvmStatic
    fun info(@Size(max = 23) tag: String, msg: String, gamStatus: GamStatus, adFormat: AdFormat, adUnitId: String) {
        log(INFO, tag, msg, gamStatus, adFormat, adUnitId)
    }

    /**
     * Prints a message with ERROR priority.
     */
    @JvmStatic
    fun error(@Size(max = 23) tag: String, msg: String, adFormat: AdFormat, adUnitId: String) {
        log(ERROR, tag, msg, GamStatus.FAILED, adFormat = adFormat, adUnitId)
    }

    /**
     * Prints a message with ERROR priority and exception.
     */
    @JvmStatic
    fun error(tag: String?, message: String?, throwable: Throwable) {
        if (tag == null || message == null) {
            return
        }
        Log.e(getTagWithBase(tag), message, throwable)
    }

    /**
     * Prints information with set priority. Every tag
     */
    private fun log(messagePriority: Int, tag: String?, message: String?, status: GamStatus, adFormat: AdFormat, adUnitId: String) {
        if (tag.isNullOrBlank() || message.isNullOrBlank()) return
        val finalTag = getTagWithBase(tag)
        Log.println(messagePriority, finalTag, message)

        // Send to server
        LogServerSender.getInstance().sendLog(
            status = status,
            message = message,
            adFormat = adFormat,
            adUnitId = adUnitId
        )
    }

    /**
     * Helper method to add Prebid tag to logging messages.
     */
    private fun getTagWithBase(tag: String): String {
        val fullTag = if (tag.startsWith(GAM_TAG)) tag else "$GAM_TAG$tag"
        return if (fullTag.length > 23) fullTag.take(22) else fullTag
    }
}